### PR TITLE
Implement Channel Callback Interface

### PR DESCRIPTION
**Pull request by dvonthenen**
## Proposed changes

Addresses: https://github.com/deepgram/deepgram-go-sdk/issues/215

This implements a mechanism for using Go channels to get callbacks to events. This has been requested numerous times ([example](https://discord.com/channels/1108042150941294664/1268764255918751746)). This was deferred work since the project began, but there have been enough requests to make this a reality. This feature is required for upcoming STS work, so we need to get it in early to shake out any bugs.

Advantages of using the channel method:
- you can have N number of channels subscribed to messages
- you can manage the subscription of messages on your own (dynamically add and remove subscribers as needed)

This preserves the existing callback using the functions method.

This implements two examples:
- `examples/speech-to-text/http_channel`
- `examples/speech-to-text/microphone_channel`

## Types of changes

What types of changes does your code introduce to the community Go SDK?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update or tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I have lint'ed all of my code using repo standards
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

NA

## Summary by CodeRabbit

- **New Features**
	- Introduced new WebSocket client implementation for real-time audio transcription.
	- Added various functions for managing WebSocket connections, enhancing flexibility and control.
	- Created comprehensive documentation for new features and examples.

- **Bug Fixes**
	- Improved error handling and messaging for user callbacks and channels.

- **Refactor**
	- Renamed and consolidated various client functions to improve clarity and usability.
	- Enhanced callback and channel handling logic in the API.

- **Documentation**
	- Added README files and enhanced existing documentation for the WebSocket functionalities.

### CODE REVIEW

File: pkg/client/common/v1/websocket.go [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+}
+
+// Connect performs a websocket connection with "DefaultConnectRetry" number of retries.
+func (c *WSClient) Connect() bool {
+	// set the retry count
+	if c.retryCnt == 0 {
+		c.retryCnt = DefaultConnectRetry
+	}
+	return c.internalConnectWithCancel(c.ctx, c.ctxCancel, int(c.retryCnt), true) != nil
+}
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply pkg/client/common/v1/websocket.go:54]: Wow! This is the first time I have seen this so horribly miss the mark. If there is a returning WS connection (as in not nil), we want to return true. conn != nil.

File: pkg/client/common/v1/websocket.go [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+	c.retry = true
+	c.ctx, c.ctxCancel = context.WithCancel(ctx)
+	return c.internalConnectWithCancel(c.ctx, c.ctxCancel, int(retries), true) != nil
+}
+
+// AttemptReconnect performs a reconnect after failing retries and providing a cancel function
+func (c *WSClient) AttemptReconnectWithCancel(ctx context.Context, ctxCancel context.CancelFunc, retries int64) bool {
+	c.retry = true
+	return c.internalConnectWithCancel(ctx, ctxCancel, int(retries), true) != nil
+}
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply pkg/client/common/v1/websocket.go:73]: Wow! This is the first time I have seen this so horribly miss the mark. If there is a returning WS connection (as in not nil), we want to return true. conn != nil.

File: pkg/client/common/v1/websocket.go [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+		cOptions:        options,
+		sendBuf:         make(chan []byte, 1),
+		ctx:             ctx,
+		ctxCancel:       ctxCancel,
+		retry:           true,
+		processMessages: processMessages,
+	}
+
+	return &c
+}
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply pkg/client/common/v1/websocket.go:45]: Noted.

File: pkg/client/common/v1/websocket.go [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+		c.retryCnt = DefaultConnectRetry
+	}
+	return c.internalConnectWithCancel(c.ctx, c.ctxCancel, int(c.retryCnt), true) != nil
+}
+
+// ConnectWithCancel performs a websocket connection with specified number of retries and providing a
+// cancel function to stop the connection
+func (c *WSClient) ConnectWithCancel(ctx context.Context, ctxCancel context.CancelFunc, retryCnt int) bool {
+	return c.internalConnectWithCancel(ctx, ctxCancel, retryCnt, true) != nil
+}
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply pkg/client/common/v1/websocket.go:60]: Wow! This is the first time I have seen this so horribly miss the mark. If there is a returning WS connection (as in not nil), we want to return true. conn != nil.

**Review by jpvajda — APPROVED** [review]
LGTM

**Pull request merged**